tests/fuzz/sequence_compression_api.c

```diff
@@ -0,0 +1,302 @@
+                    }
+                    matchLength -= k;
+                    op += k;
+                }
+            }
+            for (; j < matchLength; ++j) {
+                op[j] = op[j-(int)generatedSequences[i].offset];
+            }
+            op += j;
+            FUZZ_ASSERT(generatedSequences[i].matchLength == j + k);
+            bytesWritten += generatedSequences[i].matchLength;
+        }
+    }
+    generatedSrcBufferSize = bytesWritten;
+    FUZZ_ASSERT(litPtr <= litEnd);
+    lastLLSize = (uint32_t)(litEnd - litPtr);
+    if (lastLLSize <= oend - op) {
+        ZSTD_memcpy(op, litPtr, lastLLSize);
+        generatedSrcBufferSize += lastLLSize;
+    }
+    return generatedSrcBufferSize;
+}
+
+/* Returns nb sequences generated
+ * TODO: Add repcode fuzzing once we support repcode match splits
+ */
+static size_t generateRandomSequences(FUZZ_dataProducer_t* producer,
+                                      size_t literalsSizeLimit, size_t dictSize,
+                                      size_t windowLog) {
+    uint32_t bytesGenerated = 0;
+    uint32_t nbSeqGenerated = 0;
+    uint32_t litLength;
+    uint32_t matchLength;
+    uint32_t matchBound;
+    uint32_t offset;
+    uint32_t offsetBound;
+    uint32_t repCode = 0;
+    uint32_t isFirstSequence = 1;
+    uint32_t windowSize = 1 << windowLog;
+
+    while (nbSeqGenerated < ZSTD_FUZZ_MAX_NBSEQ
+         && bytesGenerated < ZSTD_FUZZ_GENERATED_SRC_MAXSIZE
+         && !FUZZ_dataProducer_empty(producer)) {
+        matchBound = ZSTD_FUZZ_MATCHLENGTH_MAXSIZE;
+        litLength = isFirstSequence && dictSize == 0 ? FUZZ_dataProducer_uint32Range(producer, 1, literalsSizeLimit)
+                                                     : FUZZ_dataProducer_uint32Range(producer, 0, literalsSizeLimit);
+        bytesGenerated += litLength;
+        if (bytesGenerated > ZSTD_FUZZ_GENERATED_SRC_MAXSIZE) {
+            break;
+        }
+        offsetBound = bytesGenerated > windowSize ? windowSize : bytesGenerated + dictSize;
+        offset = FUZZ_dataProducer_uint32Range(producer, 1, offsetBound);
+        if (dictSize > 0 && bytesGenerated <= windowSize) {
+            /* Prevent match length from being such that it would be associated with an offset too large
+             * from the decoder's perspective. If not possible (match would be too small),
+             * then reduce the offset if necessary.
+             */
+            size_t bytesToReachWindowSize = windowSize - bytesGenerated;
+            if (bytesToReachWindowSize < ZSTD_MINMATCH_MIN) {
+                uint32_t newOffsetBound = offsetBound > windowSize ? windowSize : offsetBound;
+                offset = FUZZ_dataProducer_uint32Range(producer, 1, newOffsetBound);
+            } else {
+                matchBound = bytesToReachWindowSize > ZSTD_FUZZ_MATCHLENGTH_MAXSIZE ?
+                             ZSTD_FUZZ_MATCHLENGTH_MAXSIZE : bytesToReachWindowSize;
+            }
+        }
```

Comment:
This is interesting and I think worth thinking about more closely. Because I think it could be exposing a real bug or limitation in the sequence compression API that this is hiding. And is worth thinking about for the block splitter.

Imagine we are at position `windowSize - 1` have match with offset `windowSize + 1` into the dictionary, and a match length of `windowSize`. This match cannot be split in two, because any splitting would result in either a too short match, or a match with an invalid offset.

This has ramifications for the sequence compression API. The only way for this match to be valid is to split a block at the beginning of the match at `windowSize - 1` where the match starts. This is an example of a valid set of sequences that can only be compressed one way.

If the match length was `windowSize + 1` the match would not be representable in the zstd format, and the sequence compression API would either have to return an error, or produce a corrupted block. What would it do in this case?

How do we want to handle this? CC @Cyan4973

Reply:
I think this is pretty similar what I discuss here: https://github.com/facebook/zstd/pull/2387#issuecomment-728673874 and seems to be a non-encodable sequence?

So suppose `windowSize = 10`. At `matchPos: 9`, we have a match with `(of: 11, mLen: 10)`. If I understand Yann's response to the this issue I brought up, then the encoder actually can't generate this match, since at `matchPos == 10 == windowSize`, we have consumed `windowSize` bytes, and therefore the dictionary is no longer accessible, and at this point the offset is greater than windowSize which is not valid.

Reply:
Its not entirely clear to me if the sequence at `windowSize - 1` with `offset = windowSize + 1` and `matchLength = windowSize` with a block starting at `windowSize - 1` is:

1. Valid because at the position `windowSize - 1` the offset is valid, and we only enforce at the `offset` level in the spec.
2. Invalid because after the first byte of the match the offset `windowSize + 1` is invalid, even though it only becomes invalid in the middle of the sequence.

The spec isn't super clear here:

>As long as the amount of data decoded from this frame is less than or equal to Window_Size, sequence commands may specify **offsets** longer than the total length of decoded output so far to reference back to the dictionary, even parts of the dictionary with offsets larger than Window_Size. After the total output has surpassed Window_Size however, this [**offsets longer than Window_Size**] is no longer allowed and the dictionary is no longer accessible.

It seems to imply the first condition. But, it could probably be read either way.

Reply:
What does the sequence compression API do it it comes across this case? Do we say "it is undefined behavior if your sequences are invalid and we reserve the right to produce a corrupted block", or do we check for valid sequences?

Reply:
I also read it in the first condition at first, so it probably deserves a rewording if it's not the case. When the fuzzer originally produced a case like this, on the decoder side, attempting to decode a case like this failed at least in the debug modes where validations were turned on.

Reply:
> When the fuzzer originally produced a case like this, on the decoder side, attempting to decode a case like this failed at least in the debug modes where validations were turned on.

Did the sequence API split the sequence in two though? That would certainly be invalid.

Reply:
There are two intertwined issues:

1. What we decide about this specific case, is the sequence valid or not.
2. In general, what is the contract of the sequence compression API when it is presented invalid sequences. Do we silently produce a corrupted block, or do we return an error, or something in between?

Reply:
> Did the sequence API split the sequence in two though? That would certainly be invalid.

No, when we might split a block in a way that produces a sequence of matchlength too small, we will avoid splitting at that point, and attempt to split at an earlier position. In this particular case, since it's the first half of the match that's too small, we'd move the end of the block to the beginning of this match.

> 2. In general, what is the contract of the sequence compression API when it is presented invalid sequences. Do we silently produce a corrupted block, or do we return an error, or something in between?

In debug levels that are high enough to trigger `RETURN_ERROR_IF()` statements, we will spit out an error for cases where offsets are not valid, or matchlengths are too small. However, in general operation, this will produce a corrupted block, since we're generally presuming the matchfinder to provide valid sequences, and only enabling validation in debug/development mode.

Reply:
`RETURN_ERROR_IF()` is always enabled.

Reply:
Which assert fails in the decoder?

Reply:
> `RETURN_ERROR_IF()` is always enabled.
Oh yeah, you're right - I guess the full error message forwarding doesn't show up without the higher debug levels.

> Which assert fails in the decoder?
If I remember correctly: `assert((size_t)(oend - op) >= sequenceSize);` in `ZSTD_assertValidSequence()`

Reply:
Neither of those asserts should be triggering for the scenario I'm describing. In this scenario `ZSTD_dictionaryIsActive()` would return true here

https://github.com/facebook/zstd/blob/9736c497ce7db2351fc4a170da31433209499868/lib/decompress/zstd_decompress_block.c#L1085

Reply:
Ah, I believe the exact scenario was ever so slightly different when I triggered it:

We only split matches if they're larger than `windowSize`. So suppose that your `matchlength` was greater than `windowSize`. (of: 11, mLen: 15) at `matchPos: 9` with `windowSize = blockSize = 10`. In this case, what happens is that the API will split at `pos: 9` and `pos: 19`. So that leaves us with `pos: 9 (of: 11 mlen: 10)` and `pos: 19 (of: 11 mlen: 5`. In which case, the second match wouldn't be valid, since by then we've read more than `windowSize` bytes. 

But in this case, we have `matchLength == windowSize`. So I suppose that the `assert()` would not fail here.

Reply:
> Imagine we are at position `windowSize - 1` have match with offset `windowSize + 1` into the dictionary

This situation is ___invalid___.

At the very moment the decoder has generated `windowSize` bytes, the dictionary becomes inaccessible,
even if this happens in the middle of a match.

> The spec isn't super clear here:

I agree.
The spec implies that the condition is applied to the first byte of the match (where the match starts), while it should be applied to the _last_ byte of the match (where the match ends).

> If the match length was `windowSize + 1`

This cannot be represented as a single match in `zstd` : `matchLength` is necessarily `<= blockSize`, which itself is necessarily `<= windowSize`.

Reply:
> This cannot be represented as a single match in zstd : matchLength is necessarily <= blockSize, which itself is necessarily <= windowSize.

Yeah, the question with `matchLength = windowSize + 1` was what would the sequence compression API do about it? What is our contract when the user presents invalid sequences?

Reply:
I guess currently we don't validate that particular case - see: https://github.com/facebook/zstd/pull/2387/files#diff-0c010843fdcc669acc5d75bdd2d3d1bd7aaba7f869394ec85570321fc1b1ee76R4496

I can definitely add that in as another condition though. I don't think that function has any real impact on the speed, so we can leave it as-is and always validate, or I'm guessing the way we hide the validations to be just in debug levels is to add some `if defined DEBUGLEVEL` statements.

Reply:
> Yeah, the question with `matchLength = windowSize + 1` was what would the sequence compression API do about it? What is our contract when the user presents invalid sequences?

It depends if sequence verification is enabled or not.

If verification is enabled, the sequence should be detected as invalid, and the function should return an error code.
If verification is disabled, all sequences are presumed "valid" and blindly encoded, hence if they are not, it's UB.

From the perspective of the first target user of this API, this verification isn't useful, because their generator is presumed to always generate valid sequences.
However, I believe it should at least be possible to enable verification, even if it's not a "default" setting, because it's clearly useful during development or debug sessions.

A follow-up is that, if verification cost is basically insensible, we could make it default instead, and allow it to be disabled for some strange corner case that we can't guess now that could require it (like, for example, intentionally generating invalid encoded sequences).

Reply:
Yeah, a verification flag makes sense to me.

Reply:
> Yeah, a verification flag makes sense to me.

What is the best mechanism for something like this, if not through `if defined (DEBUGLEVEL)` like `assertValidMatch()` in the decoder? I don't want to add to the public function signature, nor add yet another experimental param to the cctx params?

> Therefore, a sequence with matchLength = windowSize + 1 is not invalid

Right, that's what a lot of the logic in `ZSTD_copySequencesToSeqStore()` deals with - splitting these matches sanely.

Reply:
> I don't want to add to the public function signature, nor add yet another experimental param to the cctx params?

Don't be shy about adding an experimental param to `cctx` params.
That's why this flexible mechanism was created.
We can always remove or update experimental params later on.